Some suggestions for README formatting (#1)
twittemb/AsyncStateMachine: README.md

# Async State Machine

**Async State Machine** aims to provide a way to structure an application thanks to state machines. The goal is to identify the states and the side effects involved in each feature and to model them in a consistent and scalable way.

The DSL looks like this (see below for more details):

```swift
let stateMachine = StateMachine(initial: .state1) { 
    When(state: .state1) { _ in
        Execute(output: …)
    } transitions: { persons in
        On(event: …) { _ in Transition(to: .state2) }
        On(event: …) { _ in Transition(to: .state3) }
        On(event: …) { v in Transition(to: .state4(v)) }
    }
    
    When(state: …) { … in
        Execute.noOutput
    } transitions: { _ in
        …
    }
}
```

## Key points:

- Each feature is a [Moore state machine](https://en.wikipedia.org/wiki/Moore_machine): no need for a global store
- State machines are declarative: a DSL offers a natural and concise syntax
- Structured concurrency is at the core:

>    - A state machine is an AsyncSequence

Author: ```suggestion
   - A state machine is an `AsyncSequence`
```

>    - Side effects are ran inside tasks

Author: ```suggestion
   - Each side effect is run inside a `Task`
```

   - Concurrent transitions can suspend
- State machines are built in complete isolation: tests dont require mocks
- Dependencies are injected per side effect: no global bag of dependencies
- State machines are not UI related: it works with UIKit or SwiftUI

## A Simple Example

As a picture is worth a thousand words, here’s an example of a state machine that drives the opening of an elevator‘s door:

> ![](Elevator.jpeg)

Author: Instead of using a JPEG, you could take advantage of [GitHub's support](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/) for the [Mermaid DSL](https://mermaid-js.github.io/mermaid/#/flowchart) to embed charts into your Markdown, like this:

```suggestion
\```mermaid
flowchart LR
  init{" "}
  open(("open (x persons)"))
  closing(("closing (x persons)\n➡️ close(speed)"))
  closed((closed))

  init -.->|"0 persons"| open
  open --->|"personsHaveEntered(n persons)"| open
  open -->|"close button was pressed\nguard persons<10"| closing
  closing -->|"doorHasLocked"| closed
\```
```

Which would render like this:
```mermaid
flowchart LR
  init{" "}
  open(("open (x persons)"))
  closing(("closing (x persons)\n➡️ close(speed)"))
  closed((closed))

  init -.->|"0 persons"| open
  open --->|"personsHaveEntered(n persons)"| open
  open -->|"close button was pressed\nguard persons<10"| closing
  closing -->|"doorHasLocked"| closed
```

This has the advantage that your graph is also diff-able and version-controlled alongside your README. The downside is you might have a little bit less control over how it looks (even though Mermaid has a lot of customisation options, but still), but I still find it nice that this allows people to comment on the graph DSL in PRs and that the diagram is a direct part of the Markdown making it easier to keep in sync etc… so 🤷 

If you go with this solution, obviously:
 - Remove the `\` that I've put in front of the triple-backticks in my suggestion (they're only there because I can't include nested-triple-backticks into a `suggestion` block in this comment – which is already using backticks for the suggestion block — but are not meant to be part of the final syntax)
 - Don't forget to also remove the `Elevator.jpeg` image from your git repo 😉

    .map(output: Output.close(speed:), to: close(speed:))
```

> Side effects are functions that return either an `AsyncSequence` of events, or a single event. Every time the state machine produces the expected `output`, the corresponding side effect will be executed.

Author: ```suggestion
Side effects are `async` functions that return either a single `Event`, or an `AsyncSequence<Event>`. Every time the state machine produces the expected `output`, the corresponding side effect will be executed.
```
